// node-libcurl/dist/lib/enum/CurlPause.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright (c) Jonathan Cardoso Machado. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// https://github.com/curl/curl/blob/e1be8254/include/curl/curl.h#L2828
/**
 * Options to be used with `Easy.pause` or `Curl.pause`
 *
 * `CURLPAUSE_RECV_CONT` becomes `CurlPause.RecvCont`
 *
 * This also contains a special member called WriteFuncPause, that when returned from the WriteFunction callback
 *  will pause the current request
 */
var CurlPause;
(function (CurlPause) {
    CurlPause[CurlPause["Recv"] = 1] = "Recv";
    CurlPause[CurlPause["RecvCont"] = 0] = "RecvCont";
    CurlPause[CurlPause["Send"] = 4] = "Send";
    CurlPause[CurlPause["SendCont"] = 0] = "SendCont";
    CurlPause[CurlPause["All"] = 5] = "All";
    CurlPause[CurlPause["Cont"] = 0] = "Cont";
    // https://github.com/curl/curl/blob/e1be8254534898fccafc5d6cd04f6235f283cfbd/include/curl/curl.h#L252
    CurlPause[CurlPause["WriteFuncPause"] = 268435457] = "WriteFuncPause";
})(CurlPause = exports.CurlPause || (exports.CurlPause = {}));
